server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,315 @@
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// The lifetime of a handle and its resultset is tied to the lifetime of the agent.
+/// The value of the handle MUST NOT be viewed as having any meaning. Policy MUST NOT rely on
+/// its structure for any reason as it may change across releases.
+///
+/// \param[in,out] _handle       The output parameter that will hold the handle to the resultset.
+/// \param[in]     _query_string The GenQuery2 string to execute.
+/// \param[in]     _rei          This parameter is special and should be ignored.
+///
+/// \return An integer.
+/// \retval  0 On success.
+/// \retval <0 On failure.
+///
+/// \see genquery2.h for features supported by GenQuery2.
+///
+/// \since 4.3.2
+auto msi_genquery2_execute(MsParam* _handle, MsParam* _query_string, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_query_string);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_query_string->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_query_string->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_query_string->inOutStruct);
+
+    Genquery2Input input{};
+    input.query_string = static_cast<char*>(_query_string->inOutStruct);
+
+    char* results{};
+
+    if (const auto ec = rs_genquery2(_rei->rsComm, &input, &results); ec < 0) {
+        log_msi::error("{}: Error while executing GenQuery2 query [error_code=[{}]].", __func__, ec);
+        return ec;
+    }
+
+    const auto key = gq2_context.size();
+    const auto [ignored, inserted] =
+        gq2_context.insert({key, {.rows = nlohmann::json::parse(results), .current_row = -1}});
+    std::free(results); // NOLINT(cppcoreguidelines-owning-memory,cppcoreguidelines-no-malloc)
+
+    if (!inserted) {
+        log_msi::error("{}: Could not capture GenQuery2 results.", __func__);
+        return SYS_LIBRARY_ERROR;
+    }
+
+    // Return the handle to the caller.
+    fillStrInMsParam(_handle, std::to_string(key).c_str());
+
+    return 0;
+} // msi_genquery2_execute
+
+/// Moves the cursor forward by one row.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// \param[in] _handle The GenQuery2 handle.
+/// \param[in] _rei    This parameter is special and should be ignored.
+///
+/// \return An integer.
+/// \retval       0 If data can be read from the new row.
+/// \retval -408000 If the end of the resultset has been reached.
+/// \retval      <0 If an error occurred.
+///
+/// \b Example
+/// \code{.py}
+/// # A global constant representing the value which indicates the end of
+/// # the resultset has been reached.
+/// END_OF_RESULTSET = -408000;
```

Comment:
Does this need an asterisk?
```suggestion
/// *END_OF_RESULTSET = -408000;
```

Reply:
No. I learned a while back that the NREP supports constants.

Basically, you define a global, without the asterisk, outside the rules and the NREP will treat it as a constant.

Reply:
Oh okay. Then maybe the python test that looks suspiciously similar to this could drop the asterisk?

https://github.com/irods/irods/pull/7582/files#diff-cf3917e7bf5110806d6b60ae05b6da19d1998b10e351158a6e0c3e635def421aR30

Reply:
I don't know if the asterisk can be dropped there. That'll depend on how the system treats invocations via `irule`.

I've only tried using constants within the core.re file.

Reply:
Okay. I'm fine either way.

Reply:
Alrighty. Resolving.